### PR TITLE
Multiple fixes related to reading PEM key files

### LINKED ISSUE #15945: Spurious error reporting in PEM handling
This worked fine in 1.1.1 and earlier. Emits errors (but still works) in current master.

A test file that has a CSR, CERT and PRIVATE KEY within it.

This works perfectly in prior OpenSSL versions as it explicitly picks out the object that it wants in each pass.

Now we get an error message ... it still keeps going - but we should not be seeing an error condition coming out for this.

> openssl pkcs12 -inkey foo.pem -in foo.pem -export -out foo.pfx 
> 40955B1EDE7F0000:error:1608010C:STORE routines:ossl_store_handle_load_result:unsupported:crypto/store/store_result.c:151:
> 40955B1EDE7F0000:error:1608010C:STORE routines:ossl_store_handle_load_result:unsupported:crypto/store/store_result.c:151:
> 40955B1EDE7F0000:error:1608010C:STORE routines:ossl_store_handle_load_result:unsupported:crypto/store/store_result.c:151:

### PR DESCRIPTION
**Pull request by t8m**
Fixes #15945 

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
This is still incomplete

**Comment by levitte** [comment]
I'm trying to remember if we have a similar decoding loop when trying to read other objects, such as CRLs. I suppose that's the incomplete you're talking about?

File: crypto/pem/pem_pkey.c [hunk 1]
```diff
@@ -55,11 +56,23 @@ static EVP_PKEY *pem_read_bio_key_decoder(BIO *bp, EVP_PKEY **x,
...
             goto err;
-        else
+        } else {
+            if (ERR_GET_REASON(ERR_peek_error()) == ERR_R_UNSUPPORTED) {
+                /* unsupported PEM data, try again */
+                ERR_pop_to_mark();
+                ERR_set_mark();
+            } else {
+                /* other error, bail out */
+                goto err;
```
Comment (slontis) [inline crypto/pem/pem_pkey.c:71]: Wont you end up with a floating ERR_set_mark(), if you dont call ERR_clear_last_mark() here?

**Comment by t8m** [comment]
> I'm trying to remember if we have a similar decoding loop when trying to read other objects, such as CRLs. I suppose that's the incomplete you're talking about?

Yes, there is something in store as well. I'm looking at it today.

**Comment by levitte** [comment]
> Yes, there is something in store as well. I'm looking at it today.

Hold that though, I've discovered some pretty serious things in store that needs correcting.  Issue and PR coming up

File: crypto/pem/pem_pkey.c (hunk 1 above)
Comment (t8m) [reply crypto/pem/pem_pkey.c:72]: Fixup done

File: doc/man3/PEM_read_bio_PrivateKey.pod [hunk 2]
```diff
@@ -358,6 +358,10 @@ will be used.
 
 =head1 NOTES
 
+The PEM reading functions will skip any garbage or PEM data of a different
```
Comment (paulidale) [inline doc/man3/PEM_read_bio_PrivateKey.pod:361]: I'm not sure _garbage_ is great here.  How about _extraneous context_ instead?

File: crypto/pem/pem_pkey.c [hunk 3]
```diff
@@ -195,9 +209,10 @@ static EVP_PKEY *pem_read_bio_key(BIO *bp, EVP_PKEY **x,
                                   const char *propq,
                                   int selection)
 {
-    EVP_PKEY *ret;
+    EVP_PKEY *ret = NULL;
     BIO *new_bio = NULL;
     int pos;
+    struct ossl_passphrase_data_st pwdata = { 0, };
```
Comment (paulidale) [inline crypto/pem/pem_pkey.c:215]: The comma hurts my :eyes:

**Review by paulidale — APPROVED** [review]
I think there are some deep issues with how decoders and errors interact.

One wording suggestion.

**Review by paulidale — APPROVED** [review]

File: doc/man3/PEM_read_bio_PrivateKey.pod (hunk 2 above)
Comment (t8m) [reply doc/man3/PEM_read_bio_PrivateKey.pod]: fixup pushed (supposedly you've meant _extraneous content_)

File: crypto/pem/pem_pkey.c (hunk 3 above)
Comment (t8m) [reply crypto/pem/pem_pkey.c]: comma dropped, fixup pushed.

**Comment by t8m** [comment]
Merged to master. Thank you for the review.

**Pull request closed without merging**